### ISSUE TITLE
[SYCL][CUDA][HIP] query the total size of constant memory

### ISSUE THREAD

**Issue by zjin-lcf** [intel/llvm#7509]
DPCT shows the following message for the query of total size of a constant memory
"SYCL does not support the device property that would be functionally compatible with totalConstMem. "

```
cudaDeviceProp deviceProp
deviceProp.totalConstMem
```

**Comment by JackAKirk** [comment]
Interestingly it sounds like this might map to a device descriptor that is deprecated in SYCL 2020:

"
info::device::max_constant_buff
er_size:

Deprecated in SYCL 2020. Returns the maxi­
mum size in bytes of a constant buffer allo­
cation. The minimum value is 64 KB if this
SYCL device is not of type info::device_type­
::custom.
"
Although the naming, max_constant_buff
er_size, is a bit obscure: what is meant by buffer in this context? sycl::buffer?

The description of the corresponding CUDA attribute is:

cudaDevAttrTotalConstantMemory = 9
    Memory available on device for __constant__ variables in a CUDA C kernel in bytes 


Perhaps someone knows if there is a corresponding device::info for e.g. Intel devices. If it is only suitable for cuda we could add a oneapi::cuda extension for this device descriptor.

**Comment by keryell** [comment]
If the constant memory is expected to be allocated in global memory, a way could be to just return the size of the global memory instead.

**Comment by zjin-lcf** [comment]
Sorry, I don't understand your answer.

**Comment by zjin-lcf** [comment]
1.  Using CUDA constant memory can improve productivity because they (pointers) can be accessed directly in a kernel. Could SYCL allow users to migrate such CUDA programs to SYCL without passing them (pointers) as kernel arguments ?
2. If it is deprecated in 2020, then Jack proposed an extension "Perhaps someone knows if there is a corresponding device::info for e.g. Intel devices. If it is only suitable for cuda or hip we could add a oneapi::cuda/hip extension for this device descriptor."

**Comment by abagusetty** [comment]
@zjin-lcf regarding this, just forwarding some notes I have on the usage of constant memory:

There are a few options for creating a constant at program scope, depending on exactly how it’s used:

    If the value is known at host compile-time, you can declare a static constexpr variable at program scope and it can be used in device functions.
    If the value is not known until runtime, but will only be set once (e.g. some coefficients come in via the input file, but they’re fixed for the entire program run), you can use a specialization constant: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_specialization_constants. These are more powerful than __constant__ memory, because not only does the device compiler know the value is constant, it also knows what the value is – values can often be encoded directly into the instructions that use them, can be used to eliminate branches, etc.
    If the value is not known until runtime and you’ll set it multiple times (i.e. the CUDA code contains multiple calls to cudaMemcpyToSymbol or similar) then you’ll need to wait for the implementation of this “device global” extension: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_device_global.asciidoc

**Comment by zjin-lcf** [comment]
Great notes!

**Comment by zjin-lcf** [comment]
@abagusetty 
Did you observe performance improvement using specialization constant ?
